distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/JvmOptionsParser.java

```diff
@@ -0,0 +1,264 @@
+     * contents of the entire line.
+     *
+     * @param javaMajorVersion the Java major version to match JVM options against
+     * @param br the buffered reader to read line-delimited JVM options from
+     * @param jvmOptionConsumer the callback that accepts matching JVM options
+     * @param invalidLineConsumer a callback that accepts invalid JVM options
+     * @throws IOException if an I/O exception occurs reading from the buffered reader
+     */
+    static void parse(
+            final int javaMajorVersion,
+            final BufferedReader br,
+            final JvmOptionConsumer jvmOptionConsumer,
+            final InvalidLineConsumer invalidLineConsumer) throws IOException {
+        int lineNumber = 0;
+        while (true) {
+            final String line = br.readLine();
+            lineNumber++;
+            if (line == null) {
+                break;
+            }
+            if (line.startsWith("#")) {
+                // lines beginning with "#" are treated as comments
+                continue;
+            }
+            if (line.matches("\\s*")) {
+                // skip blank lines
+                continue;
+            }
+            final Matcher matcher = PATTERN.matcher(line);
+            if (matcher.matches()) {
+                final String start = matcher.group("start");
+                final String end = matcher.group("end");
+                if (start == null) {
+                    // no range present, unconditionally apply the JVM option
+                    jvmOptionConsumer.accept(line);
+                } else {
+                    final int lower;
+                    try {
+                        lower = Integer.parseInt(start);
+                    } catch (final NumberFormatException e) {
+                        invalidLineConsumer.accept(lineNumber, line);
+                        continue;
+                    }
+                    final int upper;
+                    if (matcher.group("range") == null) {
+                        // no range is present, apply the JVM option to the specified major version only
+                        upper = lower;
+                    } else if (end == null) {
+                        // a range of the form \\d+- is present, apply the JVM option to all major versions larger than the specifed one
+                        upper = Integer.MAX_VALUE;
+                    } else {
+                        // a range of the form \\d+-\\d+ is present, apply the JVM option to the specified range of major versions
+                        try {
+                            upper = Integer.parseInt(end);
+                        } catch (final NumberFormatException e) {
+                            invalidLineConsumer.accept(lineNumber, line);
+                            continue;
+                        }
+                        if (upper < lower) {
+                            invalidLineConsumer.accept(lineNumber, line);
```

Comment:
I think this should have a `continue;` after it for the sake of completeness, otherwise it's possible that it will be added to both the valid and invalid lists (it'll still fail the parsing, but to prevent future bugs)

Reply:
I pushed 919341a7261d7bdc45e186acacdb89b9e470b0d7.